File: bin/gkpd.py
#!/usr/bin/python
import os
import sys
	
file1= open("/var/lib/dpkg/info/" + sys.argv[1] + ".list")
os.system("mkdir ~/.repack/")
os.system("rm -rf ~/.repack/*")

def dosya_cek():
  data = "\n"
  while "\n" in data:
    data = file1.readline()
    j=1
    data2 = data.split("\n")
    data3 = data2[0]
    data4=data3.split("/")
    data5=""
    while j < len(data4) -1:
      data5=data5+"/"+data4[j]
      j=j+1
    
    os.system("mkdir -p ~/.repack" + data5) 
    if data3 != "":
	os.system("cp -vf " + data3 + " ~/.repack" + data3)
dosya_cek()      
os.system("mkdir ~/.repack/DEBIAN")
os.system("cp -f /var/lib/dpkg/info/" + sys.argv[1] +".* ~/.repack/DEBIAN")
os.system("cp /var/lib/dpkg/status ~/.repack/DEBIAN/status")
os.system('echo "--veri--" >> ~/.repack/DEBIAN/status')
file2= open(os.environ["HOME"]+"/.repack/DEBIAN/status")
veri = ""
arch=""
ver=""
veri = file2.readline()
while  "--veri--" not in veri:
  veri = file2.readline()
  if ("Package: " + sys.argv[1]) in veri:
    veri = veri.split("\n")
    os.system("echo "+ veri[0] + " > ~/.repack/DEBIAN/control" )
    veri = file2.readline()
    while "Package: " not in veri:
      veri = veri.split("\n")
      if "Status: " in veri[0]:
	print ""
      else:
	os.system('echo "'+ veri[0] + '" >> ~/.repack/DEBIAN/control' )
      if "Architecture:" in veri[0]:
	arch = veri[0].split(": ")[1]
      if "Version:" in veri[0]:
	ver = veri[0].split(": ")[1]
      veri = file2.readline()
    
os.system("rm -f ~/.repack/DEBIAN/*.list")
os.system("rm -f ~/.repack/DEBIAN/*.md5sums")
os.system("rm -f ~/.repack/DEBIAN/status")
os.system("mv ~/.repack/DEBIAN/*.postinst ~/.repack/DEBIAN/postinst")
os.system("mv ~/.repack/DEBIAN/*.postrm ~/.repack/DEBIAN/postrm")
os.system("mv ~/.repack/DEBIAN/*.prerm ~/.repack/DEBIAN/prerm")
os.system("mv ~/.repack/DEBIAN/*.preinst ~/.repack/DEBIAN/preinst")
os.system("mv ~/.repack/DEBIAN/*.conffiles ~/.repack/DEBIAN/conffiles")
os.system("mv ~/.repack/DEBIAN/*.shlibs ~/.repack/DEBIAN/shlibs")
os.system("mv ~/.repack/DEBIAN/*.symbols ~/.repack/DEBIAN/symbols")
os.system("mv ~/.repack/DEBIAN/*.templates ~/.repack/DEBIAN/templates")
os.system("mv ~/.repack/DEBIAN/*.config ~/.repack/DEBIAN/config")
os.system("mv ~/.repack/DEBIAN/*.triggers ~/.repack/DEBIAN/triggers")
#Baska varsa belirtin...
os.system("dpkg -b ~/.repack/")
os.system("mv  ~/.repack.deb ./"+ sys.argv[1]+"_"+ver+"_"+arch+".deb")
os.system("rm -rf ~/.repack/")

File: bin/gkpd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int dosya_cek(char *dir);
int write_control(char *name);
int make_control(char *line);
int package_file(char *name);
int copydeb(char *name);
char name[1000];
char version[50];
char arch[50];
int main(int argc, char *argv[]){
  strcpy(name,argv[1]);
  char dir[50]="/var/lib/dpkg/info/";
  strcat(dir,argv[1]);
  strcat(dir,".list");
  //open file for reading
  printf("Remove old workspace\n");
  system("rm -rf ~/.repack/ ");
  //new workspace
  printf("create new workspace\n");
  system("mkdir ~/.repack/"); 
  system("mkdir ~/.repack/DEBIAN ");
  //copy files
  int i = dosya_cek(dir);
  printf("Copy %d files\n",i);
  //create control file
  i = write_control(argv[1]);
  printf("Find %d lines\n",i);
  //copy package files
  i = package_file(argv[1]);
  printf("Copy package files\n");
  //make deb package
  printf("Make deb package: %s_%s_%s.deb\n",name,version,arch);
  system("dpkg -b ~/.repack/ 2> /dev/null | cat > /dev/null");
  printf("Move package file\n");
  i = copydeb(argv[1]);
  printf("Finish\n");
  return i;
 
}
int dosya_cek(char *dir){
   char b[50000];
   FILE *file1;
   file1 = fopen(dir,"r");
   int i = 0;
   while (fgets(b,50000,file1) != NULL )
   {
     b[strlen(b)-1] = 0;
     // cp -pvf 
     char kom1[1100]="";
     strcpy(kom1,"cp -pf ");
     strcat(kom1,b);
     strcat(kom1," ~/.repack");
     strcat(kom1,b);
     strcat(kom1," 2> /dev/null");
     char kom2[1100]="";
     int status = 1;
     char *bolu = "/";
     while (b[strlen(b)-1] != bolu[0]){
         b[strlen(b)-1]=0;
     }
     strcpy(kom2,"mkdir -p");
     strcat(kom2," ~/.repack");
     strcat(kom2,b);
     int exec = 0;
     exec = system(kom2);
     exec = system(kom1);
     i++;
   }
   return i;
}

int write_control(char *name){
   FILE *file2;
   file2 = fopen("/var/lib/dpkg/status","r");
   char c[50000];
   int d=0;
   char start[50000]="Package: "; 
   char stop[9] = "Package: ";
   strcat(start,name);
   strcat(start,"\n");
   int status = 0;
   char *a;
   while (status==0)
   {
       int j=0;
       for (j=0;j<strlen(start)+1;j++){
           if (c[j] == start[j]){
               status = 1;
        }
           else{
               status = 0;
               a= fgets(c,50000,file2);
               break;
        }
    }
}
status = 0;
while (status==0){
    d++;
    make_control(c);
    a=fgets(c,50000,file2);
    if (strstr(c,stop) != NULL)
    {
      status = 1;
    }
    else{
      status = 0;
    }    
}
   return d;
}
int make_control(char *line){
    char data[50000];
    char tmp[50000];
    if (strstr(line,"Architecture:") != NULL){
        for(int i=14;i!=strlen(line)-1;i++){
            arch[i-14]=line[i];
        }
    }
    if (strstr(line,"Version:") != NULL){
        for(int i=9;i!=strlen(line)-1;i++){
            version[i-9]=line[i];
        }
    }
    if (strstr(line,"Status:") == NULL){
        char bash[50000]="echo \"";
        line[strlen(line)-1]=0;
        strcat(bash,line);
        strcat(bash,"\" >> ~/.repack/DEBIAN/control");
        system(bash);
    }
}
int package_file(char *name){
    char bash[100]="cp -prf /var/lib/dpkg/info/";
    strcat(bash,name);
    strcat(bash,".* ~/.repack/DEBIAN/");
    system(bash);
    system("rm -f ~/.repack/DEBIAN/*.list 2> /dev/null");
    system("rm -f ~/.repack/DEBIAN/*.md5sums 2> /dev/null");
    system("rm -f ~/.repack/DEBIAN/status 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.postinst ~/.repack/DEBIAN/postinst 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.postrm ~/.repack/DEBIAN/postrm 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.prerm ~/.repack/DEBIAN/prerm 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.preinst ~/.repack/DEBIAN/preinst 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.conffiles ~/.repack/DEBIAN/conffiles 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.shlibs ~/.repack/DEBIAN/shlibs 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.symbols ~/.repack/DEBIAN/symbols 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.templates ~/.repack/DEBIAN/templates 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.config ~/.repack/DEBIAN/config 2> /dev/null");
    system("mv ~/.repack/DEBIAN/*.triggers ~/.repack/DEBIAN/triggers 2> /dev/null");
    return 0;
}
int copydeb(char *name){
 
    char bash[500] = "mv -f ~/.repack.deb $(pwd)/";
    strcat(bash,name);
    strcat(bash,"_");
    strcat(bash,version);
    strcat(bash,"_");
    strcat(bash,arch);
    strcat(bash,".deb");
    return system(bash);   
}
